### PR TITLE
Adds TS support

### PR DESCRIPTION
**Pull request by NejcZdovc**

### CODE REVIEW

**Comment by NejcZdovc** [comment]
@bbondy we have three things that we still need to do:

1) define which params are requered and which are optional
2) fix emit error
```
ERROR in [at-loader] ./test/app/background/events/runtimeEventsTest.ts:53:32
    TS2339: Property 'emit' does not exist on type 'RuntimeEvent'.

ERROR in [at-loader] ./test/app/background/events/shieldsEventsTest.ts:27:37
    TS2339: Property 'emit' does not exist on type '{ addListener: (callback: (detail: BlockDetails) => void) => void; }'.

ERROR in [at-loader] ./test/app/background/events/tabsEventsTest.ts:29:31
    TS2339: Property 'emit' does not exist on type 'TabActivatedEvent'.

ERROR in [at-loader] ./test/app/background/events/tabsEventsTest.ts:46:29
    TS2339: Property 'emit' does not exist on type 'TabCreatedEvent'.

ERROR in [at-loader] ./test/app/background/events/tabsEventsTest.ts:68:29
    TS2339: Property 'emit' does not exist on type 'TabUpdatedEvent'.

ERROR in [at-loader] ./test/app/background/events/windowsEventsTest.ts:27:37
    TS2339: Property 'emit' does not exist on type 'WindowIdEvent'.

ERROR in [at-loader] ./test/app/background/events/windowsEventsTest.ts:44:32
    TS2339: Property 'emit' does not exist on type 'WindowReferenceEvent'.

ERROR in [at-loader] ./test/app/background/events/windowsEventsTest.ts:62:32
    TS2339: Property 'emit' does not exist on type 'WindowIdEvent'.
```
3) fix one failing test 
```
  1) Shields API
       requestShieldPanelData
         resolves and calls requestShieldPanelData:
     Error: Timeout of 2000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves.
```

**Comment by NejcZdovc** [comment]
for point 3 I know what the problem is and will be resolved when point 1 is done

File: app/state/shieldsPanelState.ts [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+  return { ...state, windows }
+}
+
+export const updateFocusedWindow: shieldState.UpdateFocusedWindow = (state, windowId) =>
+  ({ ...state, currentWindowId: windowId })
+
+export const updateTabShieldsData: shieldState.UpdateTabShieldsData = (state, tabId, details) => {
+  const tabs: shieldState.Tabs = { ...state.tabs }
+
+  // TODO @bbondy do we need to add other default values? like httpsEverywhere, trackingProtection?
```
Comment (bbondy) [inline app/state/shieldsPanelState.ts]: I think these:

```
adBlock: 'allow',
trackingProtection: 'allow',
httpsEverywhere: 'allow',
javascript: 'allow'
```

File: app/types/state/shieldsPannelState.ts [hunk 2]
```diff
@@ -0,0 +1,69 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import { BlockOptions, BlockTypes } from '../other/blockTypes'
+
+// TODO  @bbondy which are required?
```
Comment (bbondy) [inline app/types/state/shieldsPannelState.ts]: I prefer all unless it's a problem, is it?

File: app/types/actions/windowActions.ts [hunk 3]
```diff
@@ -0,0 +1,43 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import * as types from '../constants/windowTypes'
+
+// TODO @bbondy what else do we have inhere?
+export interface Window {
```
Comment (bbondy) [inline app/types/actions/windowActions.ts]: You can see all properties and types here, but we only use `focused` and `id` right now.
https://developer.chrome.com/extensions/windows#type-Window

File: app/actions/webNavigationActions.ts [hunk 4]
```diff
@@ -3,7 +3,13 @@
  * You can obtain one at http://mozilla.org/MPL/2.0/. */
 
 import * as types from '../constants/webNavigationTypes'
```
Comment (bbondy) [inline app/actions/webNavigationActions.ts]: A bit confusing that this is called types now but it's ok.  Maybe we should in the future call it constants and for the TS types call it types.

File: app/background/reducers/shieldsPanelReducer.ts [hunk 5]
```diff
@@ -0,0 +1,272 @@
...
+      }
+    case windowTypes.WINDOW_REMOVED:
+      {
+        state = shieldsPanelState.removeWindowInfo(state, action.windowId)
+        break
+      }
+    case windowTypes.WINDOW_CREATED:
+      {
+        // TODO this is a dirty hack with <Window[]><any> can we improve this?
+        if (action.window.focused || (state.windows as any as Window[]).length === 0) {
```
Comment (bbondy) [inline app/background/reducers/shieldsPanelReducer.ts]: this is actually a bug, it should be Object.keys(state.windows).length === 0

File: app/types/state/shieldsPannelState.ts (hunk 2 above)
Comment (bbondy) [inline app/types/state/shieldsPannelState.ts]: I think this todo comment can be removed now

**Comment by bbondy** [comment]
Nice work!

File: app/state/shieldsPanelState.ts (hunk 1 above)
Comment (NejcZdovc) [reply app/state/shieldsPanelState.ts]: done

File: app/types/state/shieldsPannelState.ts (hunk 2 above)
Comment (NejcZdovc) [reply app/types/state/shieldsPannelState.ts]: not a problem

File: app/types/actions/windowActions.ts (hunk 3 above)
Comment (NejcZdovc) [reply app/types/actions/windowActions.ts]: done

File: app/background/reducers/shieldsPanelReducer.ts (hunk 5 above)
Comment (NejcZdovc) [reply app/background/reducers/shieldsPanelReducer.ts]: fixed

File: app/types/state/shieldsPannelState.ts (hunk 2 above)
Comment (NejcZdovc) [reply app/types/state/shieldsPannelState.ts]: removed

**Pull request merged**